File: code_snippets/ts/src/use_cases/workflows/signup_workflow.ts
import * as restate from "@restatedev/restate-sdk";

// <start_here>
// <mark_1>
const signupWorkflow = restate.workflow({
  name: "user-signup",
  handlers: {
    run: async (
      ctx: restate.WorkflowContext,
      user: { name: string; email: string }
    ) => {
      // </mark_1>
      // workflow ID = user ID; workflow runs once per user
      const userId = ctx.key;

      // <mark_2>
      await ctx.run(() => createUserEntry(user));
      // </mark_2>

      // <mark_2>
      const secret = ctx.rand.uuidv4();
      await ctx.run(() => sendEmailWithLink({ userId, user, secret }));
      // </mark_2>

      // <mark_2>
      // <mark_3>
      const clickSecret = await ctx.promise<string>("link-clicked");
      // </mark_3>
      // </mark_2>
      return clickSecret === secret;
    },

    click: async (
      ctx: restate.WorkflowSharedContext,
      request: { secret: string }
    ) => {
      // <mark_3>
      await ctx.promise<string>("link-clicked").resolve(request.secret);
      // </mark_3>
    },
  },
});
// <end_here>

export type SignUpWorkflow = typeof signupWorkflow;

restate.endpoint().bind(signupWorkflow).listen(9080);

function createUserEntry(user: { name: string; email: string }) {}

function sendEmailWithLink(param: {
  userId: string;
  user: { name: string; email: string };
  secret: string;
}) {
  return undefined;
}

File: code_snippets/python/src/use_cases/workflows/signup_workflow.py
import uuid
import restate
from pydantic import BaseModel
from restate import Workflow, WorkflowContext, WorkflowSharedContext
from src.use_cases.workflows.utils import create_user_entry, send_email_with_link

class User(BaseModel):
    name: str
    email: str


# <start_here>
user_signup = Workflow("user-signup")

# <mark_1>
@user_signup.main()
async def run(ctx: WorkflowContext, user: User) -> bool:
    # </mark_1>
    user_id = ctx.key()

    # <mark_2>
    await ctx.run("create_user", lambda: create_user_entry(user))
    # </mark_2>

    # <mark_2>
    secret = await ctx.run("secret", lambda: str(uuid.uuid4()))
    await ctx.run("send_email", lambda: send_email_with_link(user_id, user.email, secret))
    # </mark_2>

    # <mark_2>
    # <mark_3>
    click_secret = await ctx.promise("link_clicked").value()
    # </mark_3>
    # </mark_2>
    return click_secret == secret


@user_signup.handler()
async def click(ctx: WorkflowSharedContext, secret: str):
    # <mark_3>
    await ctx.promise("link_clicked").resolve(secret)
    # </mark_3>

app = restate.app(services=[user_signup])
# <end_here>